### ISSUE TITLE
kendo-angular-buttons 4.4.0 AOT broken

### ISSUE THREAD

**Issue by DanielHTpg** [telerik/kendo-angular#2250]
**Describe the bug**
I've just updated my @progress/kendo-angular-buttons package from 4.3.3 to 4.4.0 and wasn't able to successfully compile in AOT mode without warnings anymore. The application itself also wasn't running anymore in the browser. A revert to 4.3.3 fixed the issue.

I have a feeling, that this issues now comes up because of the new 4.4.0 feature "add flat ES module bundles"

**To Reproduce**
It's a local project, I haven't tested if a new project using @progress/kendo-angular-buttons 4.4.0 would also fail AOT built

**Expected behavior**
AOT compilation finished without warnings

**Screenshots**
Log of AOT compilation result:
![image](https://user-images.githubusercontent.com/24475714/57317941-86aa8300-70f9-11e9-8807-83b45f5e5385.png)

**Browser**
 - OS: Windows 10
 - Version: @progress/kendo-angular-buttons@4.4.0

**Additional context**
npm ls --depth 0 output:
```
+-- @angular-devkit/build-angular@0.13.8
+-- @angular/animations@7.2.12
+-- @angular/cli@7.3.8
+-- @angular/common@7.2.12
+-- @angular/compiler@7.2.12
+-- @angular/compiler-cli@7.2.12
+-- @angular/core@7.2.12
+-- @angular/forms@7.2.12
+-- @angular/language-service@7.2.12
+-- @angular/platform-browser@7.2.12
+-- @angular/platform-browser-dynamic@7.2.12
+-- @angular/router@7.2.12
+-- @ngrx/effects@7.4.0
+-- @ngrx/store@7.4.0
+-- @ngrx/store-devtools@7.4.0
+-- @progress/kendo-angular-buttons@4.4.0
+-- @progress/kendo-angular-charts@3.8.0
+-- @progress/kendo-angular-dateinputs@3.7.1
+-- @progress/kendo-angular-dialog@3.11.0
+-- @progress/kendo-angular-dropdowns@3.5.2
+-- @progress/kendo-angular-editor@0.4.2
+-- @progress/kendo-angular-excel-export@2.3.0
+-- @progress/kendo-angular-gauges@2.4.1
+-- @progress/kendo-angular-grid@3.13.1
+-- @progress/kendo-angular-inputs@4.2.1
+-- @progress/kendo-angular-intl@1.7.0
+-- @progress/kendo-angular-l10n@1.3.0
+-- @progress/kendo-angular-layout@3.3.0
+-- @progress/kendo-angular-messages@1.16.1
+-- @progress/kendo-angular-popup@2.5.0
+-- @progress/kendo-angular-toolbar@1.1.0
+-- @progress/kendo-angular-upload@4.3.0
+-- @progress/kendo-data-query@1.5.1
+-- @progress/kendo-drawing@1.5.10
+-- @progress/kendo-theme-material@2.4.1
+-- @types/jasmine@2.8.16
+-- @types/jasminewd2@2.0.6
+-- @types/node@8.9.5
+-- ajv@6.10.0
+-- codelyzer@4.5.0
+-- core-js@2.6.5
+-- date-fns@2.0.0-alpha.27
+-- hammerjs@2.0.8
+-- jasmine-core@2.99.1
+-- jasmine-spec-reporter@4.2.1
+-- karma@4.0.1
+-- karma-chrome-launcher@2.2.0
+-- karma-coverage-istanbul-reporter@2.0.5
+-- karma-jasmine@1.1.2
+-- karma-jasmine-html-reporter@0.2.2
+-- ngrx-store-freeze@0.2.4
+-- protractor@5.4.2
+-- rxjs@6.5.1
+-- rxjs-compat@6.5.1
+-- ts-node@7.0.1
+-- tslib@1.9.3
+-- tslint@5.15.0
+-- typescript@3.2.4
+-- web-animations-js@2.3.1
`-- zone.js@0.8.29
```

**Comment by ipeshev** [comment]
@DanielHTpg  we couldn't reproduce the issue, could you provide us a bit more information. What from @progress/kendo-angular-buttons is imported and how the project is compiled ?

**Comment by DanielHTpg** [comment]
Thanks for the replay. The application currently is using 2 lazy loaded modules for routing. I'll list the used modules in the application and their imported kendo modules. Hopefully this helps.

SharedModule:
```
const SHARED_KENDO_MODULES = [
    IntlModule,
    DialogModule,
    ToolBarModule
];

@NgModule({
    imports: [
        CommonModule,
        RouterModule,
        ...SHARED_KENDO_MODULES
    ],
    declarations: [...],
    providers: [...],
    exports: [
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        ...SHARED_KENDO_MODULES,
    ]
})
export class SharedModule { }
```

IdeasSharedModule:
```
const IDEAS_SHARED_MODULES = [
    DropDownListModule,
    InputsModule,
    GridModule,
    DateInputsModule,
    LayoutModule,
    EditorModule,
    UploadModule
];

@NgModule({
    imports: [
        SharedModule,
        ...IDEAS_SHARED_MODULES,
        ...
    ],
    declarations: [...],
    providers: [...],
    exports: [
        SharedModule,
        ...IDEAS_SHARED_MODULES,
        ...
    ],
})
export class IdeasSharedModule { }
```

IdeasModule: Lazy Loaded Route Module
```
@NgModule({
    imports: [
        IdeasRoutingModule,
        IdeasSharedModule
    ],
    declarations: [...]
})
export class IdeasModule { }
```
ChartModule: Lazy Loaded Route Module
```
@NgModule({
    imports: [
        IdeasSharedModule,
        ChartRoutingModule,
        ExcelModule,
        DropDownsModule,
        KendoChartModule,
        RadialGaugeModule,
        ...
    ],
    declarations: [ ...],
    providers: [...],
    exports: [],
})
export class ChartModule { }
```
AppModule: Entry Module
```
@NgModule({
    declarations: [
        AppComponent
    ],
    imports: [
        BrowserModule,
        BrowserAnimationsModule,
        FormsModule,
        ReactiveFormsModule,
        AppRoutingModule,
        CoreModule,
        SharedModule
    ],
    providers: [...],
    bootstrap: [AppComponent]
})
export class AppModule { }
```

AppRoutingModule :
```
const routes: Routes = [
    {
        path: '',
        redirectTo: '/ideas',
        pathMatch: 'full'
    },
    {
        path: 'ideas',
        loadChildren: './ideas/ideas.module#IdeasModule',
        canLoad: [HasAccessGuard]
    },
    {
        path: 'chart',
        loadChildren: './chart/chart.module#ChartModule',
        canLoad: [HasAccessGuard]
    },
    {
        path: 'access_denied',
        component: AccessDeniedComponent
    }
];

@NgModule({
    imports: [RouterModule.forRoot(routes)],
    exports: [RouterModule]
})
export class AppRoutingModule { }
```

**Comment by ipeshev** [comment]
Could you try to disable build-optimizer when building.

**Comment by DanielHTpg** [comment]
I've just tried it and sadly, it didn't have an effect. I ran the following command line to build:
`ng build --prod --aot --buildOptimizer=false`

**Comment by ipeshev** [comment]
@DanielHTpg you can try with [latest dev version](https://www.telerik.com/kendo-angular-ui/components/installation/development-builds/) of @progress/kendo-angular-buttons

**Comment by DanielHTpg** [comment]
I've just installed the latest version `@progress/kendo-angular-buttons@4.4.1-dev.201905141411` into the broken commit and tried comiping again but got the same errors. 

I got it working tough by "hacking" my yarn.lock file.

I've removed the previous dependency entry which is added because its a depdendency of `@progress/kendo-angular-toolbar@1.1.1:`
```
"@progress/kendo-angular-buttons@^4.3.2":
  version "4.3.3"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.3.3.tgz#a95f5cefc98b074d8b6644ab8a7f76af0ca34de5"
  integrity sha512-roJaR/k3EzbynUdFQzkkOyDvD+QKrn5Bc+cG4xNHuEp1S9VVicJDZ5VEsx9JU/zn/cZkr7aUpf0T6Wvtx/7JBg==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-popup-common" "^1.2.3"
    "@telerik/kendo-draggable" "^1.5.1"
    tslib "^1.7.0"
```
I now only have the dependency for the dev version
```
"@progress/kendo-angular-buttons@^4.4.1-dev.201905141411":
  version "4.4.1-dev.201905141411"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.4.1-dev.201905141411.tgz#8e2ed19b9481545652de8f8cad3b656f8010093e"
  integrity sha512-9k2k3Qc3o2buiVbkLX420CNkImPuug9+m2CMfgZOE3PGq08VmFP4bYPK4YuebYgDVGwHqzC1ZNWIkFwgR6HUMw==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-schematics" "^0.2.5"
    tslib "^1.7.0"
```
I have manually updated the dependency `@progress/kendo-angular-buttons` of `@progress/kendo-angular-toolbar@^1.1.1` so it looks as following:
```
"@progress/kendo-angular-toolbar@^1.1.1":
  version "1.1.1"
  resolved ".../@progress%2fkendo-angular-toolbar/-/kendo-angular-toolbar-1.1.1.tgz#083176e443bee98b8677b9d3d73a483c334bee10"
  integrity sha512-0ElYQNRzMy+jCW7GAPIgkJ0mkn1Er2Nfahs7x9SBlINMmZ80HXjkR0AWAMtOabIkGUIzxsPabaiSX5Frq0BpsA==
  dependencies:
    "@progress/kendo-angular-buttons" "^4.4.1-dev.201905141411"
    "@progress/kendo-angular-popup" "^2.0.1"
    "@progress/kendo-angular-resize-sensor" "^3.1.2"
    "@progress/kendo-schematics" "^0.2.5"
    tslib "^1.7.0"
```

I believe it's an issue in `@progress/kendo-angular-toolbar` since it uses `"@progress/kendo-angular-buttons" "^4.3.2"` as a dependency isntead of a peer dependency. Tough I'm not that sure since I'm not that familiar about the module resolution in webpack.

**Comment by tsvetomir** [comment]
@DanielHTpg that's correct - they should not be direct dependencies, but peers. Fixing it would be a breaking change, so we're a bit cautious at this point.

In any case, the 4.4.1 release will be out in a few moments. Reinstalling the modules should pick it up.

Thanks for taking the time to help us troubleshoot this issue!

**Comment by DanielHTpg** [comment]
No problem, glad I'm able to help improve an already great product 👍 

Uninstalling `@progress/kendo-angular-toolbar` and `@progress/kendo-angular-buttons`  and adding it again solves the issues. Thats apprently because at time of installation the latest version of `@progress/kendo-angular-buttons` will be resolved as dependency with the same version as when adding `@progress/kendo-angular-buttons`  as a project dependency. The dependency to peer dependency update in  `@progress/kendo-angular-toolbar` would probably make this a non issue.

After playing around with it a bit I noticed the following: 
If the resolved `@progress/kendo-angular-buttons` dependency of `@progress/kendo-angular-toolbar` is 4.3.X.X or below and the packe.json dependency `@progress/kendo-angular-buttons`  is higher than 4.4.X.X and vice versa the issue occures in AOT build. Example:
```
"@progress/kendo-angular-buttons@4.3.3":
  version "4.3.3"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.3.3.tgz#a95f5cefc98b074d8b6644ab8a7f76af0ca34de5"
  integrity sha512-roJaR/k3EzbynUdFQzkkOyDvD+QKrn5Bc+cG4xNHuEp1S9VVicJDZ5VEsx9JU/zn/cZkr7aUpf0T6Wvtx/7JBg==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-popup-common" "^1.2.3"
    "@telerik/kendo-draggable" "^1.5.1"
    tslib "^1.7.0"

"@progress/kendo-angular-buttons@^4.3.2":
  version "4.4.1"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.4.1.tgz#77ae82b416f86ab3374e67ec0862c3f1023dcac8"
  integrity sha512-2B/uVGiEePVgfq1Xly8FhYujGj/f6Ad4zU45dutBJ+mbxhazg4qo4WM7dXczlnq9PSy3N1ytSTni6MWKaNvbMA==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-schematics" "^0.2.5"
    tslib "^1.7.0"
```

Having a resolved version of 4.4.1 for `@progress/kendo-angular-buttons@^4.3.2` and a project version 4.4.0 for `@progress/kendo-angular-buttons` doesn't cause the issue. Example:

```
"@progress/kendo-angular-buttons@4.4.0":
  version "4.4.0"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.4.0.tgz#8b7a24e734baf61469359d0f4e221e2e322a09ad"
  integrity sha512-dMRvVrW40O4LvYHCRL2jx/XyGD0PQ02hGTsagCiowKOAORcwyozZgRZJAQ1/j9OoD4p+ayUnnMXdcXr2wP1awQ==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-schematics" "^0.2.5"
    tslib "^1.7.0"

"@progress/kendo-angular-buttons@^4.3.2":
  version "4.4.1"
  resolved ".../@progress%2fkendo-angular-buttons/-/kendo-angular-buttons-4.4.1.tgz#77ae82b416f86ab3374e67ec0862c3f1023dcac8"
  integrity sha512-2B/uVGiEePVgfq1Xly8FhYujGj/f6Ad4zU45dutBJ+mbxhazg4qo4WM7dXczlnq9PSy3N1ytSTni6MWKaNvbMA==
  dependencies:
    "@progress/kendo-angular-popup" "^2.0.0"
    "@progress/kendo-schematics" "^0.2.5"
    tslib "^1.7.0"
```

So aslong as the resolved minor versions of `@progress/kendo-angular-buttons` are the same, the AOT compile seems to work fine.

**Comment by tsvetomir** [comment]
With direct dependencies, npm makes no guarantees to use a single instance of the package, even when the version range is compatible. That's exactly the problem that peer dependencies solves.

The easiest workaround until this is fixed is to reinstall all modules from scratch.
```
rm -rf node_modules package-lock.json
npm install
```
